hilti/toolchain/src/compiler/global-optimizer.cc

```diff
@@ -413,20 +424,30 @@ void GlobalOptimizer::run() {
         return std::vector<NodeRef>{units.begin(), units.end()};
     }();
 
-    for ( auto& unit : units )
-        Visitor(&_hooks).collect(*unit);
+    while ( true ) {
+        bool modified = false;
 
-    for ( auto& unit : units )
-        Visitor(&_hooks).prune_uses(*unit);
+        for ( auto& unit : units )
+            Visitor(&_hooks).collect(*unit);
 
-    for ( auto& unit : units )
-        Visitor(&_hooks).prune_decls(*unit);
+        for ( auto& unit : units )
+            modified = modified || Visitor(&_hooks).prune_uses(*unit);
+
+        for ( auto& unit : units )
+            modified = modified || Visitor(&_hooks).prune_decls(*unit);
+
+        if ( ! modified )
+            break;
+
+        // After pruning the scopes might contain e.g., references
```

Comment:
Seems there's something inconsistent here about the scopes. Clearing them at the end of the loop body means that subsequent iterations will have the scopes emptied, vs. the first iteration which had them still filled. I'd say that either the collection/pruning cycle needs the scopes, then subsequent iterations now have a problem; or they don't, then we could just clear the scopes before we enter the loop.